### DISCUSSION TITLE
Is NATS officialy supported pipelining?

### DISCUSSION THREAD

**Discussion by neuecc** [nats-io/nats-server#2934] [Q&A] [answered]
Does NATS support the pipelining as described in the Redis documentation?
https://redis.io/topics/pipelining
When I tried it, messages delimited by `\r\n` could be sent in the pipeline.

`"ping\r\nsub foo.* 1\r\nsub bar.* 2\r\npub foo.try 5\r\nabcde\r\n"`

```
PONG
+OK
+OK
+OK
MSG foo.try 1 5
abcde
```

**Comment by derekcollison** [comment, marked as answer]
The protocol is agnostic to how the commands are sent, most clients batch them up to avoid excessive syscalls.

**Reply by neuecc** [reply]
Thank you.
I am currently working on another NATS client for .NET to achieve more high performance.
Thank you for your clear answer that is very helpful.

**Reply by derekcollison** [reply]
If you mean throughput performance, until io_uring on Linux becomes a main standard you still need to optimize NumMsgs/SysCall.

**Reply by neuecc** [reply]
Thanks to your advice, I have published a Nats client for .NET.
https://github.com/Cysharp/AlterNats/
